components/range_cache_memory_engine/src/range_manager.rs

```diff
@@ -199,127 +352,224 @@ impl RangeManager {
+        }
+        None
     }
 
-    pub fn pending_ranges_in_loading_contains(&self, range: &CacheRange) -> bool {
-        self.pending_ranges_loading_data
-            .iter()
-            .any(|(r, ..)| r.contains_range(range))
+    pub fn contains_region(&self, region_id: u64) -> bool {
+        self.regions.contains_key(&region_id)
     }
 
-    fn overlap_with_range(&self, range: &CacheRange) -> bool {
-        self.ranges.keys().any(|r| r.overlaps(range))
+    pub fn check_region_state(&mut self, region: &Region) -> Option<RegionState> {
+        use RegionState::*;
+        let Some(cached_meta) = self.regions.get_mut(&region.id) else {
+            return None;
+        };
+        if cached_meta.state == Pending
+            && cached_meta.region.get_region_epoch().version != region.get_region_epoch().version
+            && !cached_meta.update_region(region)
+        {
+            info!("remove outdated pending region"; "pending_region" => ?cached_meta.region, "new_region" => ?region);
+            self.remove_region(region.id);
+            return None;
+        }
+        Some(cached_meta.state)
     }
 
-    fn overlap_with_evicting_range(&self, range: &CacheRange) -> bool {
-        self.ranges_being_deleted
-            .iter()
-            .any(|(r, _)| r.overlaps(range))
+    pub fn update_region_state(&mut self, id: u64, state: RegionState) {
+        self.regions.get_mut(&id).unwrap().state = state;
     }
 
-    fn overlap_with_range_in_gc(&self, range: &CacheRange) -> bool {
-        self.ranges_in_gc.iter().any(|r| r.overlaps(range))
+    fn overlaps_with(&self, range: &CacheRange) -> bool {
+        let entry = self
+            .regions_by_range
+            .range::<[u8], (Bound<&[u8]>, Bound<&[u8]>)>((Excluded(&range.start), Unbounded))
+            .next();
+        if let Some((_key, id)) = entry {
+            let meta = &self.regions[id];
+            if meta.range.start < range.end {
+                return true;
+            }
+        }
+        false
     }
 
-    fn overlap_with_pending_range(&self, range: &CacheRange) -> bool {
-        self.pending_ranges.iter().any(|r| r.overlaps(range))
-            || self
-                .pending_ranges_loading_data
-                .iter()
-                .any(|(r, ..)| r.overlaps(range))
+    /// `check_overlap_with_region` check whether there are regions overlap with
+    /// target region. If there are regions with `pending` state and whose
+    /// epoch version is smaller than target region, the pending regions will
+    /// be removed first.
+    fn check_overlap_with_region(&mut self, region: &Region) -> Option<RegionState> {
+        let region_range = CacheRange::from_region(region);
+        let mut removed_region = vec![];
+        let mut overlapped_region_state = None;
+        self.iter_overlapped_regions(&region_range, |region_meta| {
+            if region_meta.state != RegionState::Pending
```

Comment:
maybe `if region_meta.state == RegionState::pending && region_meta.region.get_region_epoch().version < region.get_region_epoch().version { removed_region ... }` is easier to read?